jvm-streaming/src/Language/Java/Streaming.hs

```diff
@@ -0,0 +1,121 @@
+-- | Expose Java iterators as streams from the
+-- <http://hackage.haskell.org/package/streaming streaming> package.
+
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE QuasiQuotes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+{-# OPTIONS_GHC -fno-warn-unused-top-binds #-}
+
+module Language.Java.Streaming () where
+
+import Control.Distributed.Closure.TH
+import Data.IORef (IORef, newIORef, readIORef, writeIORef)
+import Data.Int (Int64)
+import Data.Singletons (SomeSing(..))
+import Data.Word (Word8)
+import Foreign.Ptr (FunPtr, castFunPtrToPtr, freeHaskellFunPtr, ptrToIntPtr)
+import qualified Foreign.JNI as JNI
+import Foreign.JNI.Types (jnull)
+import Language.Java
+import Language.Java.Inline
+import Streaming (Stream, Of)
+import qualified Streaming.Prelude as Streaming
+import System.IO.Unsafe (unsafePerformIO)
+
+isPoppableStream :: IORef (Stream (Of a) IO ()) -> IO Word8
+isPoppableStream ref = do
+    stream <- readIORef ref
+    Streaming.uncons stream >>= \case
+      Nothing -> return $ fromIntegral $ fromEnum False
+      Just _ -> return $ fromIntegral $ fromEnum True
+
+popStream :: Reflect a ty => IORef (Stream (Of a) IO ()) -> IO (J ty)
+popStream ref = do
+    stream <- readIORef ref
+    Streaming.uncons stream >>= \case
+      Nothing -> do
+        exc :: J ('Class "java.util.NoSuchElementException") <- new []
+        JNI.throw exc
+        return jnull
+      Just (x, stream') -> do
+        writeIORef ref stream'
+        reflect x
+
+type JNIFun a = JNIEnv -> JObject -> IO a
+
+foreign import ccall "wrapper" wrapObjectFun
+  :: JNIFun (J ty) -> IO (FunPtr (JNIFun (J ty)))
+foreign import ccall "wrapper" wrapBooleanFun
+  :: JNIFun Word8 -> IO (FunPtr (JNIFun Word8))
+foreign import ccall "wrapper" wrapFreeHaskellFunPtr
+  :: (FunPtr a -> IO ()) -> IO (FunPtr (FunPtr a -> IO ()))
+
+freeHaskellFunPtrPtr :: FunPtr (FunPtr a -> IO ())
+{-# NOINLINE freeHaskellFunPtrPtr #-}
+freeHaskellFunPtrPtr = unsafePerformIO (wrapFreeHaskellFunPtr freeHaskellFunPtr)
+
+newIterator :: FunPtr a -> FunPtr b -> IO (J ('Iface "java.util.Iterator"))
+newIterator hasNextPtr nextPtr = do
+    let hasNextIntPtr, nextIntPtr :: Int64
+        hasNextIntPtr = fromIntegral $ ptrToIntPtr $ castFunPtrToPtr hasNextPtr
+        nextIntPtr = fromIntegral $ ptrToIntPtr $ castFunPtrToPtr nextPtr
```

Comment:
Here this function is used only to create pointer literals in the object finalizer. It could be better to make the finalizer itself a native function, hence obviating the need for pointer literals.

That said, the quoted paragraph is essentially saying "please make sure to run this on a von Neuman architecture, not a Harvard one". We'd be hard pressed to find a plausible target on which to run inline-java that fits the latter, so the concern is pretty theoretical. In fact POSIX [mandates](http://pubs.opengroup.org/onlinepubs/9699919799/functions/V2_chap02.html#tag_15_12) that function pointers can be round tripped through pointers without loss of information. So is this something worth bothering about?